src/essential_validator.py

```diff
@@ -383,16 +383,27 @@ def validate_data(self, file_info):
             if not result:
                   return False 
                   
-        # When metadata intention is "New", all IDs must not exist in the database
-        if self.batch[BATCH_INTENTION] == INTENTION_NEW:
+        if self.batch[BATCH_INTENTION] in [INTENTION_NEW, INTENTION_DELETE]:
             ids = self.df[id_field].tolist() 
-            # query db.         
-            if not self.mongo_dao.check_metadata_ids(type, ids, self.submission_id):
-                msg = f'“{file_info[FILE_NAME]}”: duplicated data detected: “{id_field}”: {json.dumps(ids)}'
+            # query db to find existed nodes in current submission.  
+            existed_nodes = self.mongo_dao.check_metadata_ids(type, ids, self.submission_id)  
+            existed_ids = [item[NODE_ID] for item in existed_nodes]  
+            # When metadata intention is "New", all IDs must not exist in the database 
+            if len(existed_ids) > 0 and self.batch[BATCH_INTENTION] == INTENTION_NEW:
+                msg = f'“{file_info[FILE_NAME]}”: duplicated data detected: “{id_field}”: {json.dumps(duplicated_ids)}'
                 self.log.error(msg)
                 file_info[ERRORS].append(msg)
                 self.batch[ERRORS].append(msg)
                 return False
+            # When metadata intention is "Delete", all IDs must exist in the database 
+            elif self.batch[BATCH_INTENTION] == INTENTION_DELETE and len(existed_ids) < len(ids):
```

Comment:
Will it potentially fail if there are many-to-many data in the TSV? Maybe it'd be safer to remove all potential duplicated IDs before the comparison, or use sets to do comparison?

Reply:
won't.   m2m is checked in the file.   existed nodes are checked against DB.   The ids from db in the same node type and submission are unique no matter m2m relation.





[2:59](https://expanddatacommons.slack.com/archives/D05LW64MTUY/p1713812391517019)
ids from tsv file are converted to set to ensure the uniqueness.
[3:00](https://expanddatacommons.slack.com/archives/D05LW64MTUY/p1713812450308229)
not_existed_ids = list(set(ids) - set(existed_ids))

Reply:
updated.